nestkernel/conn_builder.cpp

```diff
@@ -423,6 +545,186 @@ nest::OneToOneBuilder::connect_()
+ */
+void
+nest::OneToOneBuilder::disconnect_()
+{
+  // make sure that target and source population have the same size
+  if ( sources_.size() != targets_.size() )
+  {
+    net_.message( SLIInterpreter::M_ERROR,
+      "Disconnect",
+      "Source and Target population must be of the same size." );
+    throw DimensionMismatch();
+  }
+
+#pragma omp parallel
+  {
+    // get thread id
+    const int tid = net_.get_thread_id();
+
+    try
+    {
+      // allocate pointer to thread specific random generator
+      librandom::RngPtr rng = net_.get_rng( tid );
+
+      for ( GIDCollection::const_iterator tgid = targets_.begin(), sgid = sources_.begin();
+            tgid != targets_.end();
+            ++tgid, ++sgid )
+      {
+
+        assert( sgid != sources_.end() );
+
+        if ( *sgid == *tgid and not autapses_ )
+          continue;
+
+        // check whether the target is on this mpi machine
+        if ( !net_.is_local_gid( *tgid ) )
+        {
+          skip_conn_parameter_( tid );
+          continue;
+        }
+
+        Node* const target = net_.get_node( *tgid );
+        const thread target_thread = target->get_thread();
+
+        // check whether the target is on our thread
+        if ( tid != target_thread )
+        {
+          skip_conn_parameter_( tid );
+          continue;
+        }
+        single_disconnect_( *sgid, *target, target_thread );
+      }
+    }
+    catch ( std::exception& err )
+    {
+      // We must create a new exception here, err's lifetime ends at
+      // the end of the catch block.
+      exceptions_raised_.at( tid ) =
+        lockPTR< WrappedThreadException >( new WrappedThreadException( err ) );
+    }
+  }
+}
+
+/**
+ * Solves the connection of two nodes on a OneToOne basis with
+ * structural plasticity. This means this method is used by the
+ * structural plasticity manager based on the homostatic rules defined
+ * for the synaptic elements on each node.
+ */
+void
+nest::OneToOneBuilder::sp_connect_()
+{
+  // make sure that target and source population have the same size
+  if ( sources_.size() != targets_.size() )
+  {
+    net_.message( SLIInterpreter::M_ERROR,
+      "Connect",
+      "Source and Target population must be of the same size." );
+    throw DimensionMismatch();
+  }
+
+#pragma omp parallel
+  {
+    // get thread id
+    const int tid = net_.get_thread_id();
+
+    try
+    {
+      // allocate pointer to thread specific random generator
+      librandom::RngPtr rng = net_.get_rng( tid );
+
+      for ( GIDCollection::const_iterator tgid = targets_.begin(), sgid = sources_.begin();
+            tgid != targets_.end();
+            ++tgid, ++sgid )
+      {
+        assert( sgid != sources_.end() );
+
+        if ( *sgid == *tgid and not autapses_ )
+          continue;
+
+        Node* const target = net_.get_node( *tgid );
+        const thread target_thread = target->get_thread();
+        if ( !change_connected_synaptic_elements( *sgid, *tgid, tid, 1 ) )
```

Comment:
I would suggest, that you get the `Node*` and `target_thread` after this `if`, because only then you know for sure, that the target is local. Further, I would suggest to rename the function (or split the functionality in half), so that it becomes clear, that we are also checking, if `*tgid` is a local node.

The same goes for `AllToAllBuilder::sp_connect_()`.